### PR TITLE
Audio Math: Improve audio format saturation functions.

### PR DESCRIPTION
**Pull request by ShriramShastry**
- Updated `sat_int32` to support 64-bit to 32-bit saturation with proper bitwise shifting.
- Added `vec_sat_int32x2` function to saturate and pack two 64-bit integers into `32x2` vector.
- Improved `sat_int24` for 32-bit to 24-bit saturation.
- Added `vec_sat_int24x2` to saturate and pack two 32-bit integers into a `32x2` vector.
- Corrected `sat_int16` to return a single 16-bit saturated integer.
- Added `vec_sat_int16x4` function to saturate and pack four 32-bit integers into `16x4` vector.
- Improved `sat_int8` to properly extract the lower 8 bits of a 32-bit integer.
- Added `vec_sat_int8x2` to pack two 32-bit integers in an `8x2` vector.

### CODE REVIEW

**Comment by ShriramShastry** [comment]
Performance results using Generic HiFi 3 compilers

```
						Performance Comparison
		|Function           |  Array Size  | Branching Cycles  |  HiFi Cycles      |  Bit True? |
		+-------------------+--------------+-------------------+-------------------+------------+
		|review_sat_int32   |      32      |              402  |              136  |      Yes   |
		|review_sat_int24   |      32      |              146  |               71  |      Yes   |
		|review_sat_int16   |      32      |              110  |               64  |      Yes   |
		|review_sat_int8    |      32      |              146  |               30  |      Yes   |

```

**Comment by cujomalainey** [comment]
Each line should be it's own commit, please split this. 

Please read the section on atomic comic convention [here](https://en.m.wikipedia.org/wiki/Atomic_commit#:~:text=These%20are%20sometimes%20referred%20to,focused%20nature%20of%20the%20commit.)

**Comment by ShriramShastry** [comment]
> > Performance results using Generic HiFi compilers
> > ```
> > 					Performance Comparison-hifi5:
> > 		|Function    |  Array Size  | Branching Cycles  |  HiFi Cycles      |  Bit True? |
> > 		+------------+--------------+-------------------+-------------------+------------+
> > 		|sat_int32   |      32      |              241  |              105  |      Yes   |
> > 		|sat_int24   |      32      |               77  |               43  |      Yes   |
> > 		|sat_int16   |      32      |               80  |               45  |      Yes   |
> > 		|sat_int8    |      32      |               77  |               22  |      Yes   |
> > 
> > 					Performance Comparison-hifi4:
> > 		|Function    |  Array Size  | Branching Cycles  |  HiFi Cycles      |  Bit True? |
> > 		+------------+--------------+-------------------+-------------------+------------+
> > 		|sat_int32   |      32      |              222  |              107  |      Yes   |
> > 		|sat_int24   |      32      |               75  |               41  |      Yes   |
> > 		|sat_int16   |      32      |               80  |               45  |      Yes   |
> > 		|sat_int8    |      32      |               77  |               22  |      Yes   |
> > 
> > 					Performance Comparison-hifi3
> > 		|Function    |  Array Size  | Branching Cycles  |  HiFi Cycles      |  Bit True? |
> > 		+------------+--------------+-------------------+-------------------+------------+
> > 		|sat_int32   |      32      |              402  |              136  |      Yes   |
> > 		|sat_int24   |      32      |              146  |               71  |      Yes   |
> > 		|sat_int16   |      32      |              110  |               64  |      Yes   |
> > 		|sat_int8    |      32      |              146  |               30  |      Yes   |
> > ```
> 
> @ShriramShastry can you add data that shows the before results (I think these are all "after" data). That ways its easier to see the benefits. Thanks !

Thank you; please look for the updated results table.

**Comment by lgirdwood** [comment]
@singalsu good performance updates from @ShriramShastry pls review.

**Review by lgirdwood — APPROVED** [review]
Approved the C code changes, @singalsu will need to approve the maths.

**Review by cujomalainey — APPROVED** [review]

File: src/include/sof/audio/format_hifi3.h [hunk 1]
```diff
@@ -73,6 +73,24 @@ static inline int32_t sat_int24(int32_t x)
...
+ * This function takes two 32-bit integers, packs them into a 32x2 vector,
+ * and performs saturation to 24-bit on each element.
+ */
+static inline ae_int32x2 vec_sat_int24x2(int32_t x, int32_t y)
+{
+	/* Move two 32-bit values to ae_f32x2 type */
+	ae_f32x2 d0 = AE_MOVDA32X2(x, y);
+
+	/* Saturate both values to 24-bit and pack them */
+	return (ae_int32x2)AE_SAT24S(d0);
```
Comment (singalsu) [inline src/include/sof/audio/format_hifi3.h:92]: Same concern as previous.

File: src/include/sof/audio/format_hifi3.h [hunk 2]
```diff
@@ -91,9 +91,27 @@ static inline ae_int32x2 vec_sat_int24x2(int32_t x, int32_t y)
...
+ *
+ * @param x 32-bit integer.
+ * @return 16-bit saturated integer.
+ *
+ * This function takes a 32-bit integer, packs it into a vector,
+ * and performs saturation to 16-bit on each element.
+ */
 static inline int16_t sat_int16(int32_t x)
 {
-	return AE_SAT16X4(x, x);
```
Comment (singalsu) [inline src/include/sof/audio/format_hifi3.h:27]: I don't think there's an issue with this function. Did you notice error?

Also I wonder how  more instructions added can be faster? Sure the current version moves the AE -> C type conversion to compiler but I'd not expect a difference shown by your numbers.

File: src/include/sof/audio/format_hifi3.h [hunk 3]
```diff
@@ -135,14 +135,26 @@ static inline ae_int16x4 vec_sat_int16x4(int32_t x, int32_t y, int32_t z, int32_
 	/* Saturate all values to 16-bit and pack them */
 	return AE_SAT16X4(d0, d1);
 }
+
```
Comment (singalsu) [inline src/include/sof/audio/format_hifi3.h:138]: No issues in the code but the commit text indicates an issue in the code that I don't think exists. I do agree with perf improve with this AE code version. So better to talk in commit text about optimize and not something that sounds like a bug fix.

**Review by singalsu — CHANGES_REQUESTED** [review]
I didn't see performance improvement from this with MTL xcc builds of EQs, volume, DRC, TDFB. But also there wasn't any degrade of performance, same results within a fraction of MCPS. I think it's because most of the code is sat_int32() that wasn't improved (at least much) by looks of the code though you reported some improve. Can you check especially the s24 case.

File: src/include/sof/audio/format_hifi3.h [hunk 4]
```diff
@@ -15,9 +15,22 @@
...
+ * @param x 64-bit integer.
+ * @return 32-bit saturated integer.
+ *
+ * This function takes a 64-bit integer, performs bitwise shifting,
+ * and returns a saturated 32-bit integer.
+ */
 static inline int32_t sat_int32(int64_t x)
 {
-	return (ae_int32)AE_ROUND32F48SSYM(AE_SLAI64(x, 16));
+	/* Shift left by 32 bits with saturation */
```
[the remark this answers is not in the document]
Comment (singalsu) [reply src/include/sof/audio/format_hifi3.h:29]: Also, do you have idea why your version is faster in your test?

**Comment by lgirdwood** [comment]
> I didn't see performance improvement from this with MTL xcc builds of EQs, volume, DRC, TDFB. But also there wasn't any degrade of performance, same results within a fraction of MCPS. I think it's because most of the code is sat_int32() that wasn't improved (at least much) by looks of the code though you reported some improve. Can you check especially the s24 case.

Its probably difficult to see the gain from logs if its hidden in larger infra especially as we as saving single instructions in some cases. If logs are marginal AND maths is correct then we should trust the tooling that these are valid improvements and merge as future use cases may hit these APIs harder where benefits in the logs will be more obvious.

**Comment by singalsu** [comment]
> > I didn't see performance improvement from this with MTL xcc builds of EQs, volume, DRC, TDFB. But also there wasn't any degrade of performance, same results within a fraction of MCPS. I think it's because most of the code is sat_int32() that wasn't improved (at least much) by looks of the code though you reported some improve. Can you check especially the s24 case.
> 
> Its probably difficult to see the gain from logs if its hidden in larger infra especially as we as saving single instructions in some cases. If logs are marginal AND maths is correct then we should trust the tooling that these are valid improvements and merge as future use cases may hit these APIs harder where benefits in the logs will be more obvious.

There is no answers to my questions yet, not removing my reject. I have especially concern about 24 bit saturate intrinsics compatibility with the tricky S24_4LE format in SOF. I can check it today myself of course.

File: src/include/sof/audio/format_hifi3.h [hunk 5]
```diff
@@ -1,40 +1,179 @@
...
+	ae_f64 shifted = AE_SLAI64S(x, 32);
+
+	/* Shift right by 32 bits */
+	return (int)AE_MOVINT32_FROMINT64(AE_SRAI64(shifted, 32));
```
Comment (singalsu) [inline src/include/sof/audio/format_hifi3.h:33]: the typecast should be (int32_t).

File: src/include/sof/audio/format_hifi3.h (hunk 2 above)
Comment (ShriramShastry) [reply src/include/sof/audio/format_hifi3.h:27]: Performance Analysis
Performance Comparison
The performance comparisons between the original and the revised implementations show identical results:
Original GitHub Code:
```
`w/o` loop vectorization

Function           | Array Size | Branching Cycles | HiFi Cycles | Bit True? 
-------------------+------------+------------------+-------------+-----------
sat_int16          |     32     |         80       |      73     |    Yes   

`w` loop vectorization
Function           | Array Size | Branching Cycles | HiFi Cycles | Bit True? 
-------------------+------------+------------------+-------------+-----------
sat_int16          |     32     |         80       |      45     |    Yes   

```

Pull Request Code:

```
`w/o` loop vectorization

Function           | Array Size | Branching Cycles | HiFi Cycles | Bit True? 
-------------------+------------+------------------+-------------+-----------
sat_int16          |     32     |         80       |      73     |    Yes   

`w` loop vectorization
Function           | Array Size | Branching Cycles | HiFi Cycles | Bit True? 
-------------------+------------+------------------+-------------+-----------
sat_int16          |     32     |         80       |      45     |    Yes   
```

Error Checking and Performance Analysis
• The function has no explicit errors and produces expected bit-true outcomes.

Instruction Count and Performance
• The intrinsic `AE_MOVDA32` copies a 32-bit input into both the high and low halves of a 32x2 output.
• The `AE_SAT16X4` conducts four-way saturation of 32-bit inputs to 16-bit values.
• The `AE_MOVAD16_0` function pulls a 16-bit value from the packed results.
• Performance findings are consistent (HiFi Cycles: 73).
• Instruction differences effect code readability and clarity while maintaining efficient performance.
Both the original and revised implementations achieve the same performance levels with the same bit-true results

`w/o` loop vectorization

```
void dsp_write_loop_int16_bitwise(int16_t* arr, size_t size) {
    for (size_t i = 0; i < size; i += 1) {
        arr[i] = review_sat_int16(arr[i]);
    }
}

```
`w` loop vectorization

```
void dsp_write_loop_int16_bitwise(int16_t* arr, size_t size) {
    ae_int16x4* ptr_out = (ae_int16x4*)arr;
    int reminder = size % 4;
    int i;

    for (i = 0; i < size / 4; i += 1) {
        size_t base_idx = i * 4;
        ptr_out[i] = review_sat_int16x4(arr[base_idx], arr[base_idx + 1], arr[base_idx + 2], arr[base_idx + 3]);
    }

    i = (size / 4) * 4;
    for (; i < size; i += 1) {
        arr[i] = review_sat_int16(arr[i]);
    }
}
```

**Review by ShriramShastry — COMMENTED** [review]
Thank you so much for examining the code. I've included my responses. Please have a look.

File: src/include/sof/audio/format_hifi3.h [hunk 6]
```diff
@@ -33,6 +33,27 @@ static inline int32_t sat_int32(int64_t x)
...
+ * This function takes two 64-bit integers, converts them to ae_f64 format,
+ * and returns them as a packed ae_int32x2 vector after saturation.
+ */
+static inline ae_int32x2 vec_sat_int32x2(int64_t x, int64_t y)
+{
+	ae_f64 d0, d1;
+
+	/* Convert 64-bit integers to ae_f64 format */
+	d0 = int64_rtor_ae_f64(x);
+	d1 = int64_rtor_ae_f64(y);
```
[the remark this answers is not in the document]
Comment (singalsu) [reply src/include/sof/audio/format_hifi3.h:52]: What is the "Branching Cycles" and why does it differ for these ways? Is (ae_f64) typecast better with smaller number? Still the HiFi cycles are the same. Not critical but if typecast way is good for you, I prefer it.

File: src/include/sof/audio/format_hifi3.h (hunk 1 above)
Comment (singalsu) [reply src/include/sof/audio/format_hifi3.h:92]: No concern, this code is correct.

File: src/include/sof/audio/format_hifi3.h (hunk 2 above)
Comment (singalsu) [reply src/include/sof/audio/format_hifi3.h:27]: I meant that your commit text "Corrected to return ..." indicates that there would be some issue. Just reformulate the commit text.

Can you summarize if you proposed version is better MCPS than existing? Just yes or no :) If there are other merits vs. original please add to commit text.

**Review by singalsu — COMMENTED** [review]
Thanks for responses, many resolved issues. Some still left, mainly commit text improvement need I think

**Review by ShriramShastry — COMMENTED** [review]
I have address the review comments. Please take look, 
Thank you.

**Comment by lgirdwood** [comment]
@singalsu good for you ?

File: src/include/sof/audio/format_hifi3.h [hunk 7]
```diff
@@ -53,9 +53,23 @@ static inline ae_int32x2 vec_sat_int32x2(int64_t x, int64_t y)
 	/* Round and saturate both 64-bit values to 32-bit and pack them */
 	return (ae_int32x2)AE_ROUND32X2F64SSYM(d0, d1);
 }
+
```
Comment (singalsu) [inline src/include/sof/audio/format_hifi3.h:56]: Nit about commit text, the new version isn't more accurate, it's mote efficient.

**Review by singalsu — APPROVED** [review]
